packages/pyright-internal/src/tests/samples/dataclass26.py

```diff
@@ -0,0 +1,54 @@
+# This sample tests assignment of dataclass fields that use
+# the coverter parameter described in PEP 712.
+
+from dataclasses import dataclass, field
+from typing import Generic, TypeVar
+
+
+def converter_simple(s: str) -> int:
+    return int(s)
+
+
+@dataclass
+class Foo:
+    # This should generate an error because "converter" is not an official property yet.
+    field0: int = field(converter=converter_simple)
+
+foo = Foo("1")
+reveal_type(foo.field0, expected_text="int")
+foo.field0 = "2"
+
+# This should generate an error because only strs can be assigned to field0.
```

Comment:
You might want to add a test case that verifies that type narrowing doesn't occur when assigning to a field with a converter. To do so, just copy the above `reveal_type` line after the `foo.field0 = "2"` assignment statement. It should still be considered an `int`, whereas normal narrowing rules would cause it to become a `str`.

Reply:
Another interesting test is the case where the converter doesn't actually do any conversion (i.e. it accepts the same type as it produces). That shouldn't be considered an asymmetric descriptor, so normal type narrowing rules would apply. For example, if the converter accepts a `str | int` and produces a `str | int` and you assign a value of "1" to the field, it should be narrowed to `Literal[1]`.